File: Hard/2147. Number of Ways to Divide a Long Corridor.java
class Solution {
    private static final int MOD = (int) 1e9 + 7;

    public int numberOfWays(String corridor) {
        long ans = 1;
        int cntS = 0, cntP = 0;
        for (char c : corridor.toCharArray()) {
            if (c == 'S') {
                if (cntS > 0 && cntS % 2 == 0) {
                    ans = (ans * (cntP + 1)) % MOD;
                    cntP = 0;
                }
                cntS++;
            } else {
                if (cntS > 0 && cntS % 2 == 0) {
                    cntP++;
                }
            }
        }
        return (cntS > 0 && cntS % 2 == 0) ? (int) ans : 0;
    }
}

File: Hard/2147. Number of Ways to Divide a Long Corridor.py
MOD = 1000000007
class Solution:
    def numberOfWays(self, corridor: str) -> int:
        seats = [i for i, v in enumerate(corridor) if v == 'S']
        if not seats or len(seats) % 2:
            return 0
        ways = 1
        for i in range(2, len(seats), 2):
            plants = seats[i] - seats[i - 1]
            ways *= plants
            ways %= MOD
        return ways

class Solution1:
    def numberOfWays(self, corridor: str) -> int:
        s = p = 0
        ways = 1
        for c in corridor:
            if c == 'S':
                if s and s % 2 == 0:
                    ways *= p + 1
                    p = 0
                s += 1
            else:
                if s and s % 2 == 0:
                    p += 1
        return ways % 1000000007 if s and s % 2 == 0 else 0